File: Top 50 Coding Interview Questions/09-Write a code for bubble sort/BubbleSort.py
# Function to perform Bubble Sort
def bubble_sort(arr):
    n = len(arr)

    # Outer loop for each pass through the array
    for i in range(n - 1):
        # Flag to check if any swapping occurred in this pass
        swapped = False

        # Inner loop for comparing adjacent elements
        for j in range(n - i - 1):
            # Swap if the current element is greater than the next element
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True  # Mark as swapped

        # If no swapping occurred, the array is already sorted
        if not swapped:
            break

# Test the Bubble Sort function
arr = [64, 34, 25, 12, 22, 11, 90]
print("Original array:", arr)

bubble_sort(arr)

print("Sorted array:", arr)

File: Top 50 Coding Interview Questions/09-Write a code for bubble sort/bubblesort.c
#include<stdio.h>

//Function to print array
void display(int arr[], int size){
    int i;
    for(i = 0; i<size;i++)
    {
        printf("%d  ", arr[i]);
    }
    printf("\n");
}
//Main function to run the program
int main(){
    int array[]={5,3,1,9,8,2,4,7,10};
    int size = sizeof(array) / sizeof(array[0]);

    printf("Before bubble sort: \n");
    display(array,size);

    int i, j ,temp;
    for(i = 0; i<size-1;i++){
        //since,after each iteration rightmost in element are sorted
        for(j = 0; j < size - i -1; j++)
        if(array[j]>array[j + 1]){
            temp = array[j]; //swap the element
            array[j] = array[j + 1];
            array[j + 1] = temp;
        }
    }
    printf("After bubble sort: \n");
    display(array ,size);
    return 0;
}